src/net/client/xfr.rs

```diff
@@ -0,0 +1,1123 @@
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum XfrEvent<R> {
+    /// Delete record R in zone serial S.
+    ///
+    /// The transfer signalled that the given record should be deleted from
+    /// the zone version with the given serial number.
+    ///
+    /// Note: If the transfer contains N deletions of fhe same record then
+    /// this event will occur N times.
+    DeleteRecord(Serial, R),
+
+    /// Add record R in zone serial S.
+    ///
+    /// The transfer signalled that the given record should be added to the
+    /// zone version with the given serial number.
+    ///
+    /// Note: If the transfer contains N additions of fhe same record then
+    /// this event will occur N times.
+    AddRecord(Serial, R),
+
+    /// Prepare to delete records in zone serial S.
+    ///
+    /// The transfer signalled that zero or more record deletions will follow,
+    /// all for the zone version with the given serial number.
+    BeginBatchDelete(Serial),
+
+    /// Prepare to add records in zone serial S.
+    ///
+    /// The transfer signalled that zero or more record additions will follow,
+    /// all for the zone version with the given serial number.
+    BeginBatchAdd(Serial),
+
+    /// Transfer completed successfully.
+    ///
+    /// Note: This event is not emitted until the final record of the final
+    /// response in a set of one or more transfer responss has been seen.
+    EndOfTransfer,
+
+    /// Transfer processing failed.
+    ///
+    /// This event indicates that there is a problem with the transfer data
+    /// and that transfer processing cannot continue.
+    ProcessingFailed,
+}
+
+//--- Display
+
+impl<R> std::fmt::Display for XfrEvent<R> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            XfrEvent::DeleteRecord(_, _) => f.write_str("DeleteRecord"),
+            XfrEvent::AddRecord(_, _) => f.write_str("AddRecord"),
+            XfrEvent::BeginBatchDelete(_) => f.write_str("BeginBatchDelete"),
+            XfrEvent::BeginBatchAdd(_) => f.write_str("BeginBatchAdd"),
+            XfrEvent::EndOfTransfer => f.write_str("EndOfTransfer"),
+            XfrEvent::ProcessingFailed => f.write_str("ProcessingFailed"),
+        }
+    }
+}
+
+//------------ XfrEventHandler ---------------------------------------------------
+
+/// A trait for implementing handlers of [`XfrEvent`]s.
+pub trait XfrEventHandler {
+    /// Handle the given [`XfrEvent`].
+    ///
+    /// Returning an Err will cause transfer processsing to be aborted and the
+    /// error to be returned to the client of [`XfrResponseProcessor`], except in
+    /// the case of [`XfrEvent::ProcessingFailed`] for which the return value of
+    /// this handler will be ignored by [`XfrResponseProcessor`].
+    fn handle_event(
+        &self,
+        evt: XfrEvent<XfrRecord>,
+    ) -> impl std::future::Future<Output = Result<(), Error>> + Send;
```

Comment:
I’m not sure we should use impl trait in trait method return types just yet. I think at this time I would prefer if we use an associated type for the future type and provide a blanket impl for `Fn(...)` that uses boxed trait objects.

Reply:
The Rust compiler told me to do it this way ;-) I _think_ I understand your suggestion, will see if I actually do.

Reply:
By the way, if I understand correctly, the blanket Fn impl approach if the same as the one we had previously for `service()` kept causing problems with (a) not being able to implement other impls due to overlapping impls and (b) caused very confusing compiler errors when the compiler thought I was trying in client code to impl the Fn trait but actually I wasn't. So, if that's what you mean, I'm not so sure it's a good idea.

Reply:
Maybe we can avoid the explicit use of futures by flipping this on its head? The processor could return some sort of iterator over events for each answer given to it.

Reply:
This is still implemented as a Future but now as an associated type to work around minimal versions CI error  `"impl Trait only allowed in function and inherent method return types, not in trait method return"`.

Reply:
Following internal discussions the design has now been changed to a sync iterator-per-response approach. I will resolve this outdated comment.